src/euler_one_qubit_decomposer.rs

```diff
@@ -10,15 +10,635 @@
+        lam += PI;
+        theta = -theta;
+        phi += PI;
+    }
+    lam = mod_2pi(lam);
+    if lam.abs() > atol {
+        global_phase += lam / 2.;
+        circuit.push((String::from(k_gate), vec![lam]));
+    }
+    circuit.push((String::from(a_gate), vec![theta]));
+    phi = mod_2pi(phi);
+    if phi.abs() > atol {
+        global_phase += phi / 2.;
+        circuit.push((String::from(k_gate), vec![phi]));
+    }
+    OneQubitGateSequence {
+        gates: circuit,
+        global_phase,
+    }
+}
+
+fn circuit_u3(
+    theta: f64,
+    phi: f64,
+    lam: f64,
+    phase: f64,
+    simplify: bool,
+    atol: Option<f64>,
+) -> OneQubitGateSequence {
+    let mut circuit = Vec::new();
+    let phi = mod_2pi(phi);
+    let lam = mod_2pi(lam);
+    let atol = match atol {
+        Some(atol) => atol,
+        None => DEFAULT_ATOL,
+    };
+    if !simplify || theta.abs() > atol || phi.abs() > atol || lam.abs() > atol {
+        circuit.push((String::from("u3"), vec![theta, phi, lam]));
+    }
+    OneQubitGateSequence {
+        gates: circuit,
+        global_phase: phase,
+    }
+}
+
+fn circuit_u321(
+    theta: f64,
+    phi: f64,
+    lam: f64,
+    phase: f64,
+    simplify: bool,
+    atol: Option<f64>,
+) -> OneQubitGateSequence {
+    let mut circuit = Vec::new();
+    let mut atol = match atol {
+        Some(atol) => atol,
+        None => DEFAULT_ATOL,
+    };
+    if !simplify {
+        atol = -1.0;
+    }
+    if theta.abs() < atol {
+        let tot = mod_2pi(phi + lam);
+        if tot.abs() > atol {
+            circuit.push((String::from("u1"), vec![tot]));
+        }
+    } else if (theta - PI / 2.).abs() < atol {
+        circuit.push((String::from("u2"), vec![mod_2pi(phi), mod_2pi(lam)]));
+    } else {
+        circuit.push((String::from("u3"), vec![theta, mod_2pi(phi), mod_2pi(lam)]));
+    }
+    OneQubitGateSequence {
+        gates: circuit,
+        global_phase: phase,
+    }
+}
+
+fn circuit_u(
+    theta: f64,
+    phi: f64,
+    lam: f64,
+    phase: f64,
+    simplify: bool,
+    atol: Option<f64>,
+) -> OneQubitGateSequence {
+    let mut circuit = Vec::new();
+    let mut atol = match atol {
+        Some(atol) => atol,
+        None => DEFAULT_ATOL,
+    };
+    if !simplify {
+        atol = -1.0;
+    }
+    let phi = mod_2pi(phi);
+    let lam = mod_2pi(lam);
+    if theta.abs() > atol || phi.abs() > atol || lam.abs() > atol {
+        circuit.push((String::from("u"), vec![theta, phi, lam]));
+    }
+    OneQubitGateSequence {
+        gates: circuit,
+        global_phase: phase,
+    }
+}
+
+fn circuit_psx_gen<F, P, X>(
+    theta: f64,
+    phi: f64,
+    lam: f64,
+    phase: f64,
+    simplify: bool,
+    atol: Option<f64>,
+    mut pfun: P,
+    mut xfun: F,
+    xpifun: Option<X>,
+) -> OneQubitGateSequence
+where
+    F: FnMut(&mut OneQubitGateSequence),
+    P: FnMut(&mut OneQubitGateSequence, f64),
+    X: FnMut(&mut OneQubitGateSequence),
```

Comment:
Strictly I think `X` is eligible to be `FnOnce`, but I can't imagine it matters much since this isn't a public API.

Reply:
I can try it and see it wouldn't hurt to make it more restrictive.

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/9578/commits/5745c830b2464c9d3761f6a55bbc5475fcbab8a6